File: Sources/App/Services/Phisherman.swift
import Foundation
import Vapor

struct Phisherman {

// check function that returns data
  func check(domain: String) async -> {

    guardDomain(domain)

    let domain = domain.canonicalizeURL()

    let apiKey = ProcessInfo.processInfo.environment["PHISHERMAN_API_KEY"]

    let url = URL(
      string: "https://api.phisherman.gg/v2/domains/check/\(domain)")!

    var request = URLRequest(url: url)
    request.httpMethod = "GET"
    request.setValue("internal-server@phish.directory", forHTTPHeaderField: "User-Agent")
    request.setValue("internal-server@phish.directory", forHTTPHeaderField: "X-Identity")
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    request.setValue("https://phish.directory", forHTTPHeaderField: "Referer")

    // add bearer token
    request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")

    // make the request
    let (data, _) = try await URLSession.shared.data(for: request)

    let response = try JSONDecoder().decode(PhishermanResponse.self, from: data)

    return response
  }
}

File: src/services/Phisherman.ts
import axios from "axios";

import { getDbDomain } from "../functions/db/getDbDomain";
import { prisma } from "../prisma";

/**
 * A service that provides access to the Phisherman service for checking and reporting domains.
 */
export class PhishermanService {
  domain = {
    /**
     * Asynchronously checks a given domain against the Phisherman service for any known bad domains.
     *
     * @param {string} domain - The domain name to be checked.
     * @returns
     */
    check: async (domain: string) => {
      // metrics.increment("services.phisherman.domain.check");

      const response = await axios.get(
        `https://api.phisherman.gg/v2/domains/info/${domain}`,
        {
          // todo: extract headers to a seperate place to avoid duplication (will need to handle adding the Authorization header)
          headers: {
            Authorization: "Bearer " + process.env.PHISHERMAN_API_KEY!,
            Referer: "https://phish.directory",
            "User-Agent": "internal-server@phish.directory",
            "X-Identity": "internal-server@phish.directory",
          },
        }
      );

      const data = response.data;
      const dbDomain = await getDbDomain(domain);

      await prisma.rawAPIData.create({
        data: {
          sourceAPI: "Phisherman",
          domain: {
            connect: {
              id: dbDomain.id,
            },
          },
          data: data,
        },
      });

      return data;
    },
  };
}
